### ISSUE TITLE
Integrate new parser to support pdftotree output

### ISSUE THREAD

**Issue by lukehsiao** [HazyResearch/fonduer#3]
TODO:
- [x] Create simple, verifiable test data for unit testing the parser
- [ ] Get numbers from old parser for comparison
- [ ] Test current version of new parser
- [ ] Rewrite to fix the mismatches. That is, build the document model, and run each paragraph through spacy for phrases, rather than chunking the whole document like was done for CoreNLP.

**Comment by dbulhosa** [comment]
Hey @lukehsiao! I tried using different PDF to HTML converters (pdfminer.six, pdf2htmlEX, pdftotree) and by far the converter that best preserved the semantic structure of the PDF text when parsed was pdftotree. However, when using pdftotree I get the error: `Document not added to database because of parse error: "list index out of range"`. 

I am going to try using Acrobat Reader but my ultimate goal is to create an application using Fonduer, so I need to be able to do the conversion programmatically. As I understand pdftotree is not maintained. Is there another recommended PDF to HTML programmatic converter outside of pdftotree?

**Comment by lukehsiao** [comment]
> As I understand pdftotree is not maintained.

Yes, that's correct. But it is "unmaintained" in the sense there is no active development right now. If you wanted to track down that bug an open a PR, we'd be happy to review and merge it. 

> Is there another recommended PDF to HTML programmatic converter outside of pdftotree? Incidentally, at least for my PDFs, pdftotree seems to do a better job than Acrobat -- would be great to see it incorporated into Fonduer!

Our experience is similar to yours. We tried a whole lot of different packages (ABBYY is another fairly popular option you didn't mention), but non work great. That was the motivation for starting pdftotree, but we don't have the resources to work on it more at the moment, and Acrobat, with it's flaws, seems to do OK for the datasets we've encountered. I will mention that Adobe does have "Actions", so you can set up a pipeline of steps and apply it to a whole directory. That what we usually use so it is at least somewhat automated.

One caveat is that in our experience, their Actions can leak memory (and eventually crash), so if you set one up, you should set it up to do something like take a PDF, convert it to HTML, and move the PDF file to a different directory. That way you can find of pick up where you left off if/when Acrobat crashes.

**Comment by dbulhosa** [comment]
Hey! I really appreciate you getting back to me. I may try tracking down the bug if I have the time for it and need to. I did try the Acrobat actions and it worked great for bulk converting PDFs but like you mention I got some crashes. In the end I don't think I'll be able to use it because there isn't a license + SDK for the type of application I need to create.

I am finding though that the output of pdf2htmlEX is a good starting point. The output of it does not cause errors. Furthermore, the classes it output by it roughly discretize the positions of different spans, which is a really useful feature. It also seems that the styling is reduced to as few classes as possible. This means that by comparing position-adjacent spans with similar styling I can do a decent job of merging tags correctly to preserve the semantics of the text and get the right sentences. The position discretization is not perfect though, so I may go back and do the merging based on the actual position. 

The cool thing though is that if I combine,

- `pdf2htmlEx` with `--heps=200px`.
- The `premailer` module in Python to move the `pdf2htmlEx` generated styling from separate style tags to inline in the tags containing the text.
- My custom preprocessing routine to merge tags based on position adjacency and styling match based on the `pdf2htmlEx` class outputs.
- The `tinycss` module in Python to parse style attributes in html tags

I can use the class "style clusters" created by `pdf2htmlEX` as well as the detailed style properties from `premailer` to create throttlers and label functions in Fonduer. I'm going to keep going this route and see how well it works out. Could be a neat alternative.
